crates/libcontainer/src/process/fork.rs

```diff
@@ -1,37 +1,77 @@
+// process. This is used to launch the container init process so the parent
+// process of the calling process can receive ownership of the process. If we
+// clone a child process as the init process, the calling process (likely the
+// youki main process) will exit and the init process will be re-parented to the
+// process 1 (system init process), which is not the right behavior of what we
+// look for.
+pub fn container_clone_sibling<F: FnOnce() -> Result<i32>>(cb: F) -> Result<Pid> {
+    let mut clone = clone3::Clone3::default();
+    let mut pidfd = -1;
+    // Note: normally, an exit signal is required, but when using
+    // `CLONE_PARENT`, the `clone3` will return EINVAL if an exit signal is set.
+    // The older `clone` will not return EINVAL in this case. Instead it ignores
+    // the exit signal bits in the glibc wrapper.
+    clone.flag_pidfd(&mut pidfd).flag_parent();
+
+    container_clone(cb, clone).with_context(|| "failed to clone sibling process")
+}
+
+// A simple clone wrapper to clone3 so we can share this logic in different
+// fork/clone situations. We decided to minimally support kernel version >= 5.4,
+// and `clone3` requires only kernel version >= 5.3. Therefore, we don't need to
+// fall back to `clone` or `fork`.
+fn container_clone<F: FnOnce() -> Result<i32>>(
+    cb: F,
+    mut clone_cmd: clone3::Clone3,
+) -> Result<Pid> {
+    // Return the child's pid in case of parent/calling process, and for the
+    // cloned process, run the callback function, and exit with the same exit
+    // code returned by the callback. If there was any error when trying to run
+    // callback, exit with -1
+    match unsafe { clone_cmd.call().with_context(|| "failed to run clone3")? } {
+        0 => {
+            // Inside the cloned process
             let ret = match cb() {
                 Err(error) => {
-                    log::debug!("failed to run fork: {:?}", error);
+                    log::debug!("failed to run child process in clone: {:?}", error);
                     -1
                 }
                 Ok(exit_code) => exit_code,
             };
             std::process::exit(ret);
         }
+        pid => Ok(Pid::from_raw(pid)),
     }
 }
 
+// Execute the cb in another process. Make the fork works more like thread_spawn
+// or clone, so it is easier to reason. Compared to clone call, fork is easier
+// to use since fork will magically take care of all the variable copying. If
+// using clone, we would have to manually make sure all the variables are
+// correctly send to the new process, especially Rust borrow checker will be a
+// lot of hassel to deal with every details.
+pub fn container_fork<F: FnOnce() -> Result<i32>>(cb: F) -> Result<Pid> {
+    // Using `clone3` to mimic the effect of `fork`.
+    let mut clone = clone3::Clone3::default();
+    let mut pidfd = -1;
+    clone.flag_pidfd(&mut pidfd).exit_signal(SIGCHLD as u64);
```

Comment:
Do you have the plan to use pidfd?

Reply:
I don't see a need for this, so I can remove them for now, until we need it.